### ISSUE TITLE
limits of this module

### ISSUE THREAD

**Issue by kintaro1981** [Real-Gecko/virtualmin-nginx-fpm#11]
Hello,

I would like to know what limits have this module.

It is designed for a single virtual server or is possibile to manage multiple virtual servers each one with a different configuration? It support sub-servers?

Regards

**Comment by Real-Gecko** [comment]
Hi, I use module to manage all virtual servers on my production server.
However it's functionality not really extensive.
For now it only creates Nginx and FPM configs:

```
server {
    server_name domain.tld www.domain.tld;
    listen 127.0.0.1;

    root /home/domain_owner/public_html;
    index index.html index.htm index.php;
    access_log /var/log/nginx/domain.tld_access_log;
    error_log /var/log/nginx/domain.tld_error_log;
    fastcgi_param GATEWAY_INTERFACE CGI/1.1;
    fastcgi_param SERVER_SOFTWARE nginx;
    fastcgi_param QUERY_STRING $query_string;
    fastcgi_param REQUEST_METHOD $request_method;
    fastcgi_param CONTENT_TYPE $content_type;
    fastcgi_param CONTENT_LENGTH $content_length;
    fastcgi_param SCRIPT_FILENAME /home/domain_owner/public_html$fastcgi_script_name;
    fastcgi_param SCRIPT_NAME $fastcgi_script_name;
    fastcgi_param REQUEST_URI $request_uri;
    fastcgi_param DOCUMENT_URI $document_uri;
    fastcgi_param DOCUMENT_ROOT /home/domain_owner/public_html;
    fastcgi_param SERVER_PROTOCOL $server_protocol;
    fastcgi_param REMOTE_ADDR $remote_addr;
    fastcgi_param REMOTE_PORT $remote_port;
    fastcgi_param SERVER_ADDR $server_addr;
    fastcgi_param SERVER_PORT $server_port;
    fastcgi_param SERVER_NAME $server_name;
    fastcgi_param HTTPS $https;
    location ~ (^|/)\. {
        return 403;
    }
    location / {
        try_files $uri $uri/ /index.php?q=$uri&$args;
    }
    location ~ \.php$ {
        try_files $uri =404;
        fastcgi_pass unix:/var/run/php5-fpm-domain.tld.sock;
    }
}
```

```
user = domain_owner
group = domain_owner
listen = /var/run/php5-fpm-domain.tld.sock
listen.owner = www-data
listen.group = www-data
pm = dynamic
pm.max_children = 1
pm.start_servers = 1
pm.min_spare_servers = 1
pm.max_spare_servers = 1
pm.max_requests = 50
chdir = /
env[TMP] = /home/domain_owner/.tmp
env[TMPDIR] = /home/domain_owner/.tmp
env[TEMP] = /home/domain_owner/.tmp
php_admin_value[cgi.fix_pathinfo] = 0
php_admin_value[open_basedir] = /home/domain_owner/public_html/:/home/domain_owner/.tmp/
php_admin_value[session.save_path] = /home/domain_owner/.tmp/
```

Theese settings are configurable through Virtualmin interface.

```
pm = dynamic
pm.max_children = 1
pm.start_servers = 1
pm.min_spare_servers = 1
pm.max_spare_servers = 1
pm.max_requests = 50
```

You can create sub-servers also, for now that's all.If you wish to have some more features, create new issues.

**Comment by kintaro1981** [comment]
so, in case I want to personalized the configuration of specific virtual server the way to go is modify configuration file in site-enabled, right? I can't do it trought Virtualmin UI, right?

**Comment by Real-Gecko** [comment]
What config are you asking about exactly? Nginx one or FPM?

**Comment by kintaro1981** [comment]
well... as said before, I'm totally newbie to nginx 😄 
Can I assume that FPM configs rarely are different from one virtualserver to another, right?
So I'm speaking of nginx's configs.

I totally need to study it a little more. 😉

**Comment by Real-Gecko** [comment]
Well, Nginx config is not changable and is created as written above, this pattern was taken from the core Nginx module (actually copied from already created config by the core module) and works pretty fine. The FPM itself does not offer much configuration, so actually the only tning you can tweak with Virtualmin interface is performance - to lower memory usage on less loaded sites, or increase it on high load sites.
If you need some tweaks in Nginx config then create new issues in this repo and I'll see what can I do to improve this module.

**Comment by 7starsone** [comment]
Hello, some questions please:
if I enable this plugin, does it automatically apply to the existing virtual servers and how?
Log rotation still not working?
Can I use opcache per user...and one only instance per user?
With FPM is really avoidable the fcgid max processes issue as in mod fcgid? That is, are resources always available for the users...so better managed? thanks

**Comment by Real-Gecko** [comment]
There is a "readme" in [this repo root](https://github.com/Real-Gecko/virtualmin-nginx-fpm) describing how to install and use this plugin.
Log rotation is not still working.
Opcache is installed globally in the system, this plugin does not interfere with it in any manner.
You can change:
```
- FPM control type 	
- Number of child processes 	
- The number of child processes to create on startup 	
- The desired minimum number of idle processes 	
- The desired maximum number of idle processes 	
- The number of requests each child process should execute before respawning
```

**Comment by 7starsone** [comment]
sorry, the readme doesn't explain what happens to the existing vhosts...  Do I need to go into any vhost, one by one, and in Enabled Features, instead of "Apache website enabled?" there is something about Nginx or that will be automatically enabled?

**Comment by Real-Gecko** [comment]
Yes, that's correct, you'll have to switch your vhosts manually. Note that this plugin is not compatible with stock Apache plugin, but plays nicely with stock Nginx plugin.

**Comment by kintaro1981** [comment]
watch out @7starsone of the other issues because in some cases if you touch specific option it will reset the configuration to the default one.

**Comment by Real-Gecko** [comment]
Yep, plugin definitely needs rework :D

**Comment by kintaro1981** [comment]
Please @Real-Gecko find some time to fix that. I think that many virtualmin user will be happy about that. (I'm thinking to leave Virtualmin as it seam that they will not add official support anytime soon).

**Comment by Real-Gecko** [comment]
If only Webmin/Virtualmin API would be a little bit saner, it would be a piece of cake :D

**Comment by 7starsone** [comment]
yes, indeed...if this plugin was done good, could be great for any virtualmin user... there's really the lack of an easy Php-fpm management...it's really odd 'cause fpm is officially out since 5 years and even before was in development...by now is known by a too long time to ignore....

**Comment by Real-Gecko** [comment]
By the way, did anyone tried it with PHP7?
I'm curious how PHP-FPM plays with new version of PHP...

**Comment by 7starsone** [comment]
try to talk to andreychek on irc channel and Jamie....maybe working together is quicker...

**Comment by kintaro1981** [comment]
I found this now (support FPM too):
https://github.com/TwinFuture/virtualmin-nginx-ssl/
Maybe you can join efforts. 👍

**Comment by Real-Gecko** [comment]
Hmm, never seen this plugin, what features does it have?

**Comment by kintaro1981** [comment]
I just found it, it was added only 25 days ago.

**Comment by Real-Gecko** [comment]
Cool, maybe it'll become a better alternative for my plugin :D

**Comment by kintaro1981** [comment]
I'm going to ask some info. Let's see.

https://github.com/TwinFuture/virtualmin-nginx-ssl/issues/1

**Comment by 7starsone** [comment]
also the non ssl version https://github.com/TwinFuture/virtualmin-nginx

**Comment by kintaro1981** [comment]
yep

**Comment by kintaro1981** [comment]
@7starsone @Real-Gecko take a look here:
https://www.virtualmin.com/comment/768102#comment-768102

**Comment by Real-Gecko** [comment]
Nice

**Comment by kintaro1981** [comment]
@Real-Gecko 
with virtualmin 5.06 the support of php-fpm is official, there are only some bugs that are going to be fixed in the next release:
https://www.virtualmin.com/project/issues?text=php-fpm&projects=&status=Open&categories=All&order=created&sort=desc

**Comment by Real-Gecko** [comment]
I tried official FPM on one of my domains and got `Bad Gateway` eventhough created configs looked correct. Anyway this feature looks really promissing. As soon as official FPM support will become 100% workable I'll drop support of this module.
